core/internal/api/api.go

```diff
@@ -106,6 +106,20 @@ type Request struct {
 	Headers map[string]string
 }
 
+func (req *Request) String() string {
+	s := "Request{"
+	s += "Method: " + req.Method + ", "
+	s += "Path: " + req.Path + ", "
+	s += "Body: " + string(req.Body) + ", "
+	s += "Headers: {"
+	for k, v := range req.Headers {
+		s += k + ": " + v + ", "
+	}
+	s += "}"
+	s += "}"
+	return s
+}
```

Comment:
`strings.Builder` would be slightly more efficient here. It could matter since we always construct the string even if logging is disabled; but the main reason to do it is style---string concatenation in a loop is a smell.

Though the neatest, relatively-efficient way would be `fmt.Sprintf`:

```go
fmt.Sprintf(
    "Request{Method: %s, Path: %s, Body: %s, Headers: %v}",
    req.Method,
    req.Path,
    string(req.Body),
    req.Headers,
)
```

This'll print Headers in the format `map[Header:Value Header:Value ...]`, which I think is OK.